### PR TITLE
[IMP] discuss: support for voice attachments

### PR DESCRIPTION
**Pull request by aktr-odoo**
This commit supports processing voice attachments while processing attachments for posts. It checks for voice information when handling attachments and includes voice-specific data appropriately. Attachments flagged with voice are now correctly processed after being added and voice metadata will be created for them. This will enhance the system’s ability to manage voice-related attachments.

Task-3839838

### CODE REVIEW

File: addons/mail/models/mail_thread.py [hunk 1]
```diff
@@ -2334,15 +2336,18 @@ def _process_attachments_for_post(self, attachments, attachment_ids, message_val
...
             new_attachments = self.env['ir.attachment'].sudo().create(attachement_values_list)
             attach_cid_mapping, attach_name_mapping = {}, {}
-            for attachment, (cid, name, token) in zip(new_attachments, attachement_extra_list):
+            for attachment, (cid, name, token, voice) in zip(new_attachments, attachement_extra_list):
                 if cid:
                     attach_cid_mapping[cid] = (attachment.id, token)
                 if name:
                     attach_name_mapping[name] = (attachment.id, token)
+                if voice:
+                    attachment._post_add_create(voice=True)
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_thread.py:2350]: This is going to call "register as main attachment", although it is already managed. I would instead create discuss.voice.metadata manually (in batch) here and avoid calling "do a bit of everything" method as we are in an optimized processus here.

Comment (tde-banana-odoo) [inline addons/mail/models/mail_thread.py:2350]: Test should be added in community for that also

**Review by tde-banana-odoo — COMMENTED** [review]
Zboing, sorry for the delay, some remarks :)

**Comment by aktr-odoo** [comment]
Hello @tde-banana-odoo 

Sorry for the delay, The changes have been made can you please have a look?

File: addons/mail/models/discuss/discuss_channel.py [hunk 2]
```diff
@@ -633,6 +633,26 @@ def _message_post_after_hook(self, message, msg_vals):
...
+                    if voice:
+                        # Extract attachment_id from the (4, attachment_id) M2M command
+                        attachment_id = m2m_command[1] if m2m_command[0] == 4 else None
+                        if attachment_id:
+                            voice_metadata_vals.append({
+                                'attachment_id': attachment_id,
+                            })
+            if voice_metadata_vals:
+                self.env['discuss.voice.metadata'].create(voice_metadata_vals)
+        return return_values
```
Comment (seb-odoo) [inline addons/mail/models/discuss/discuss_channel.py:654]: Hello,

Isn't it supposed to be done in `_post_add_create`?

The voice kwargs might not be correctly passed (see recent fix https://github.com/odoo/odoo/pull/184506)
or the method might not be called in the flow you are trying to fix?

Comment (aktr-odoo) [reply addons/mail/models/discuss/discuss_channel.py:654]: Hello @seb-odoo 

The `_post_add_create` isn't triggered in the flow.

Maybe we can do,

```py
# Extract attachment_id from the (4, attachment_id) M2M command
attachment_id = m2m_command[1] if m2m_command[0] == 4 else None
if attachment_id:
    self.env['ir.attachment'].browse(attachment_id).exists()._post_add_create(voice=True)
```

What say?

**Comment by aktr-odoo** [comment]
Hello @seb-odoo 

Thanks for the review.

The requested changes have been made. Can you please have a look?

File: addons/mail/models/discuss/discuss_channel.py (hunk 2 above)
Comment (tde-banana-odoo) [reply addons/mail/models/discuss/discuss_channel.py]: If you look at the docstring: "Overrides behaviour when the attachment is created through the controller"

File: addons/mail/models/discuss/discuss_channel.py [hunk 3]
```diff
@@ -633,6 +633,21 @@ def _message_post_after_hook(self, message, msg_vals):
...
+        if attachments:
+            for attachment, m2m_command in zip(attachments, return_values.get('attachment_ids', [])):
+                if len(attachment) == 3:
+                    _, _, info = attachment
+                    voice = info and info.get('voice')
+                    if voice:
+                        # Extract attachment_id from the (4, attachment_id) M2M command
+                        attachment_id = m2m_command[1] if m2m_command[0] == 4 else None
+                        if attachment_id:
+                            self.env['ir.attachment'].browse(attachment_id).exists()._post_add_create(voice=True)
```
Comment (seb-odoo) [inline addons/mail/models/discuss/discuss_channel.py:648]: Looking for closely, I don't think it is the right place to put the `_post_add_create`.

If the attachments are already created, the "voice" (through `_post_add_create`) should be set at their creation, not when they are linked to a message.

Comment (aktr-odoo) [reply addons/mail/models/discuss/discuss_channel.py:648]: The attachments are created in `_process_attachments_for_post` in the `mail.thread` but as the voice is coming from the discuss channel we have created a voice here.

Comment (aktr-odoo) [reply addons/mail/models/discuss/discuss_channel.py:648]: > Where is the info voice key coming from originally?

The voice is coming form the whatsapp side with in he webhook. See https://github.com/odoo/enterprise/pull/63075
Whenever user is sending audio message from Whatsapp, we will receive voice in webhook.
In the flow, we will extract information from the webhook and will post it in active channel using `message_post`.
image

**Comment by tde-banana-odoo** [comment]
As we target stable, I made https://github.com/odoo/odoo/pull/186244 which is simpler. We cannot call "post add create" as it is targetting "upload a file", not posting with attachments. The issue lies in post add create, not in current code.

Cheers !

**Pull request closed without merging**